src/Generating/FinishGen.cpp

```diff
@@ -801,6 +805,11 @@ void cFinishGenSprinkleFoliage::GenFinish(cChunkDesc & a_ChunkDesc)
 					)
 					{
 						a_ChunkDesc.SetBlockType(x, ++Top, z, E_BLOCK_CACTUS);
+
+						for (int i = 0; i < (x % m_MaxCactusHeight); i++)
+						{
+							a_ChunkDesc.SetBlockType(x, Top + i + 1, z, E_BLOCK_CACTUS);
```

Comment:
You do need to modify the Top variable because it's used to set the chunk's heightmap.  This will leave the heightmap wrong.  Same with the sugarcane.